### PR TITLE
New API and Manta functionality

### PR DESCRIPTION
**Pull request by cheapRoc**
Ref: #9 #19 

Here's the final API re-organization as well as the introduction of @jen20's `manta-go`.

Recap: The re-organization effort came about from the need to merge two code bases in a single repo (for manta) while respecting existing functionality. It then made sense to take the moment to simplify and refactor each package, grouping around cloud resources, and considering Joyent branding.

The old repo is pretty much deprecated and (hopefully) this will be the basis moving forward.

This is a behemoth of a PR which is something I don't normally like do. Most of the commits are file moves/renames. I took time to ensure implementations were kept the same. I believe `client/`, `storage/`, and `compute/`, contain the most changes.

## Features

- Each package contains it's own client.
- Client now accepts a shared `triton.ClientConfig` input struct.
- MANTA_URL can be configured at the Client (along with CloudAPI endpoint).
- `manta-go` functions now respect the `context.Context` package.
- Almost all methods have been renamed and/or simplified given their package name.
- Input structs pretty much stay the same.
- `Machines` => `Instances`
- There is still work to clean-up the `client` package.

## Example

Here's a working example of the new API.

<details>

```go
package main

import (
	"context"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"time"

	triton "github.com/joyent/triton-go"
	"github.com/joyent/triton-go/account"
	"github.com/joyent/triton-go/authentication"
	"github.com/joyent/triton-go/compute"
	"github.com/joyent/triton-go/identity"
	"github.com/joyent/triton-go/network"
	"github.com/joyent/triton-go/storage"
)

func main() {
	keyID := os.Getenv("SDC_KEY_ID")
	accountName := os.Getenv("SDC_ACCOUNT")
	keyPath := os.Getenv("SDC_KEY_FILE")

	privateKey, err := ioutil.ReadFile(keyPath)
	if err != nil {
		log.Fatal(err)
	}

	sshKeySigner, err := authentication.NewPrivateKeySigner(keyID, privateKey, accountName)
	if err != nil {
		log.Fatal(err)
	}

	config := &triton.ClientConfig{
		Endpoint:    os.Getenv("SDC_URL"),
		MantaURL:    os.Getenv("MANTA_URL"),
		AccountName: accountName,
		Signers:     []authentication.Signer{sshKeySigner},
	}

	// create Account client
	a, err := account.NewClient(config)
	if err != nil {
		log.Fatalf("account.NewClient: %s", err)
	}

	// account.Config ----------------------------------------------------------

	cfg, err := a.Config().Get(context.Background(), &account.GetConfigInput{})
	if err != nil {
		log.Fatalf("account.Config.Get: %v", err)
	}

	acct, err := a.Get(context.Background(), &account.GetInput{})
	if err != nil {
		log.Fatalf("account.Get: %v", err)
	}

	// identity.Roles.List -----------------------------------------------------

	i, err := identity.NewClient(config)
	if err != nil {
		log.Fatalf("identity.NewClient: %s", err)
	}

	roles, err := i.Roles().List(context.Background(), &identity.ListRolesInput{})
	if err != nil {
		log.Fatalf("ListRoles died with %v", err)
	}
	numRoles := 0
	for _, role := range roles {
		fmt.Println("** Role:", role.Name)
		numRoles++
	}
	if numRoles == 0 {
		fmt.Println("No roles found")
	}

	// compute.Instances.List --------------------------------------------------

	c, err := compute.NewClient(config)
	if err != nil {
		log.Fatalf("compute.NewClient: %s", err)
	}

	listInput := &compute.ListInstancesInput{}
	instances, err := c.Instances().List(context.Background(), listInput)
	if err != nil {
		log.Fatalf("compute.Instances.List: %v", err)
	}
	numInstances := 0
	for _, instance := range instances {
		numInstances++
		fmt.Println(fmt.Sprintf("-- Instance: %v", instance.Name))
	}

	// network.ListNetworks ----------------------------------------------------

	n, err := network.NewClient(config)
	if err != nil {
		log.Fatalf("network.NewClient: %s", err)
	}

	nets, err := n.List(context.Background(), &network.ListInput{})
	if err != nil {
		log.Fatalf("network.List: %v", err)
	}
	numNets := 0
	for _, net := range nets {
		fmt.Println("++ Network:", net.Name)
		numNets++
	}

	// storage.Objects.Put -----------------------------------------------------

	licensePath := "/Users/justinreagor/go/src/github.com/joyent/triton-go/LICENSE"
	licenseReader, err := os.Open(licensePath)
	if err != nil {
		log.Fatalf("os.Open: %s", err)
	}
	defer licenseReader.Close()

	s, err := storage.NewClient(config)
	if err != nil {
		log.Fatalf("storage.NewClient: %s", err)
	}

	stamp := time.Now().Format("030405")
	objectPath := fmt.Sprintf("/stor/LICENSE-%s", stamp)
	putInput := &storage.PutObjectInput{
		ObjectPath:   objectPath,
		ObjectReader: licenseReader,
	}
	err = s.Objects().Put(context.Background(), putInput)
	if err != nil {
		log.Fatalf("PutObject died with: %s", err)
	}

	// storage.Dir.List --------------------------------------------------------

	listInput2 := &storage.ListDirectoryInput{
		DirectoryName: "/stor",
	}
	dir, err := s.Dir().List(context.Background(), listInput2)
	if err != nil {
		log.Fatalf("ListMachines died with %v", err)
	}
	numEntries := 0
	for _, entry := range dir.Entries {
		fmt.Println(entry.Name)
		numEntries++
	}

	// -------------------------------------------------------------------------

	fmt.Println("Account Login:", acct.Login)
	fmt.Println("Default Network:", cfg.DefaultNetwork)
	fmt.Println("Total Networks: ", numNets)
}
```

</details>

### CODE REVIEW

**Comment by cheapRoc** [comment]
Also /cc @tgross @sean-

**Comment by cheapRoc** [comment]
One change I thought of while working through Terraform was error messaging. I believe my replaces might have cobbled stuff like `Error executing CreateMachine request` => `Error executing Create request`. I'll revert these back to reflect their CloudAPI endpoint names.

**Comment by tgross** [comment]
So just a sanity check on the process here: the examples / tests don't seem to work anymore. There aren't any commits touching the README either, which suggests that it probably needs updating too. Maybe we could drop that great extended example from the top-level comment in this PR into the README?

```
# GOPATH=/Users/tim.gross/src/joyent/triton-go
# SDC_URL=https://us-east-1.api.joyent.com
# SDC_ACCOUNT=timgross
# SDC_KEY_ID=redacted

$ go get -x github.com/joyent/triton-go
...
$ TRITON_TEST=1 go test -v ./...
storage/examples/createjob/main.go:9:2: cannot find package "github.com/jen20/manta-go" in any of:
        /usr/local/Cellar/go/1.8/libexec/src/github.com/jen20/manta-go (from $GOROOT)
        /Users/tim.gross/src/joyent/triton-go/src/github.com/jen20/manta-go (from $GOPATH)
storage/examples/createjob/main.go:10:2: cannot find package "github.com/jen20/manta-go/authentication" in any of:
        /usr/local/Cellar/go/1.8/libexec/src/github.com/jen20/manta-go/authentication (from $GOROOT)
        /Users/tim.gross/src/joyent/triton-go/src/github.com/jen20/manta-go/authentication (from $GOPATH)
storage/examples/mls/main.go:6:2: cannot find package "github.com/davecgh/go-spew/spew" in any of:
        /usr/local/Cellar/go/1.8/libexec/src/github.com/davecgh/go-spew/spew (from $GOROOT)
        /Users/tim.gross/src/joyent/triton-go/src/github.com/davecgh/go-spew/spew (from $GOPATH)

**Comment by cheapRoc** [comment]
Yes, thanks for the reminder, the README.md, examples, and tests, definitely need to be updated along with this change set. I completely overlooked that I included `manta-go`'s old `storage/examples/` in this. I can dust those off as well.

**Comment by cheapRoc** [comment]
Probably makes sense to create an `examples/` directory straight off root since we don't have formal unit testing yet.

**Comment by gwydirsam** [comment]
Didn't see this work happening over here. Checkout https://github.com/jen20/manta-go/pull/6 -- it fixes the error codes.

**Comment by gwydirsam** [comment]
I'm going to go ahead and move examples to the root and update them for the new API.

**Comment by cheapRoc** [comment]
Nice catch @gwydirsam! We should probably include your changes after we close this PR. There's a few other refinements after this change set anyway.

Just to clarify, besides your removal of `respBody.Close()` within `GetObject`, it looks like your changes to error messages should easily port over into the new [client/errors.go](https://github.com/joyent/triton-go/pull/23/files#diff-1fc53dfcba5f976f893a21390f12e8ee) file. Correct?

File: triton.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+// Universal package used for defining configuration used across all client
+// constructors.
+
+// ClientConfig is a placeholder/input struct around the behavior of configuring
+// a client constructor through the implementation's runtime environment
+// (SDC/MANTA env vars).
+type ClientConfig struct {
+	Endpoint    string
```
Comment (jen20) [inline triton.go]: Should we rename this to `TritonURL` to mirror `MantaURL`? Since this would be breaking, now would be the best time to do this.

Comment (gwydirsam) [inline triton.go]: This. This caused a bug in signedURLs that I fixed here: https://github.com/gwydirsam/triton-go/commit/1740705ff9a9fe9912cb33fc84acb07595d49153

**Comment by cheapRoc** [comment]
Everything sounds reasonable. There are definitely optimizations to be made after the PR is finalized, i.e. client package, unit tests, etc.

I've got a couple patches to get the tests functional again. I'll post up some changes later today.

**Comment by cheapRoc** [comment]
Looks like the tests have diverged from how they were originally setup. Prior to recent changes, the tests utilized a global client interface. Each package now has its own client so bootstrapping tests will need to take this into account somehow.

**Comment by cheapRoc** [comment]
Had some free time to focus on what it would take to fix the existing acceptance tests. Here's the API change I'm working on now. It's still early but I believe this will help exercise every top level client and package.

I've introduced `testutils.StepClient` which exposes how each client is initialized. The global `triton.ClientConfig` is passed into this handler much the same way as we passed along the original global client.

<details>

```go
func TestAccDataCenters_Get(t *testing.T) {
	const dataCenterName = "us-east-1"
	const dataCenterURL = "https://us-east-1.api.joyentcloud.com"

	testutils.AccTest(t, testutils.TestCase{
		Steps: []testutils.Step{
			&testutils.StepClient{
				StateBagKey: "datacenter",
				CallFunc: func(config triton.ClientConfig) (interface{}, error) {
					client, err := compute.NewClient(config)
					if err != nil {
						log.Fatalf("compute.NewClient: %s", err)
					}
				},
			},

			&testutils.StepAPICall{
				StateBagKey: "datacenter",
				CallFunc: func(client *compute.ComputeClient) (interface{}, error) {
					ctx := context.Background()
					input := &compute.GetInput{
						Name: dataCenterName,
					}
					return client.Datacenters().Get(ctx, input)
				},
			},

			&testutils.StepAssert{
				StateBagKey: "datacenter",
				Assertions: seq.Map{
					"name": dataCenterName,
					"url":  dataCenterURL,
				},
			},
		},
	})
}
```

</details>
<br>
Redundancy does exist in this approach in that we'd need to do this for every test. Only other option would be to add an internal switch within the runner which pairs every unique `StateBagKey` with a matching client init step.

**Comment by cheapRoc** [comment]
I believe those last commits bring back the full functionality of the acceptance tests. I'll have free time on the plane Sunday to work on examples and docs.

**Comment by cheapRoc** [comment]
Also @jen20, I only tested under my own account, not `tritongo`. It didn't look like my key was setup on that account.

**Comment by gwydirsam** [comment]
@cheapRoc If you have an operator account you can use your key and just hit the tritongo account

**Comment by cheapRoc** [comment]
Feeling confident I've wrapped this one up.

/cc @jen20 @sean- @tgross

**Comment by cheapRoc** [comment]
I should note, I'll expand into the examples a bit more after the PR. I'd also like to refactor the storage job output function as well, its not outputting jobs properly at the moment.

**Comment by cheapRoc** [comment]
Ignore that last commit, looks like I didn't merge master but picked the PR itself. 👎

**Comment by jen20** [comment]
@cheapRoc I think we're good to merge here now!

**Comment by cheapRoc** [comment]
:+1: one epic PR closed

**Comment by misterbisson** [comment]
👍 nice work all

**Pull request merged**